File: python/traderdaddy/backoff.py
"""backoff.py — Exponential backoff with jitter for rate-limited calls.

Port of @traderdaddy/sdk's backoff.ts (base 2s, cap 60s, +1s jitter). Retries
only on ``RateLimitError``; any other error propagates on the first throw.
"""

from __future__ import annotations

import asyncio
import random
from typing import Awaitable, Callable, Optional, TypeVar

from .errors import RateLimitError

T = TypeVar("T")

RetryCallback = Callable[[int, float, RateLimitError], None]
Sleep = Callable[[float], Awaitable[None]]


def backoff_delay_ms(
    attempt: int,
    *,
    base_ms: float = 2_000,
    cap_ms: float = 60_000,
    jitter_ms: float = 1_000,
) -> float:
    """Jittered delay for a given 0-based attempt index."""
    delay = min(base_ms * (2**attempt), cap_ms)
    return delay + random.random() * jitter_ms


async def with_backoff(
    fn: Callable[[], Awaitable[T]],
    *,
    retries: int = 4,
    base_ms: float = 2_000,
    cap_ms: float = 60_000,
    jitter_ms: float = 1_000,
    on_retry: Optional[RetryCallback] = None,
    sleep: Optional[Sleep] = None,
) -> T:
    """Run ``fn``, retrying on ``RateLimitError`` with exponential backoff.

    Re-throws the last ``RateLimitError`` once ``retries`` is exhausted; any
    non-rate-limit error is raised immediately.
    """
    _sleep: Sleep = sleep or (lambda ms: asyncio.sleep(ms / 1000))
    attempt = 0
    while True:
        try:
            return await fn()
        except RateLimitError as err:
            if attempt >= retries:
                raise
            delay = backoff_delay_ms(
                attempt, base_ms=base_ms, cap_ms=cap_ms, jitter_ms=jitter_ms
            )
            if on_retry is not None:
                on_retry(attempt + 1, delay, err)
            await _sleep(delay)
            attempt += 1

File: src/backoff.ts
/**
 * backoff.ts — Exponential backoff with jitter for rate-limited calls.
 *
 * Ported from DaddyBoard's `poller.js` backoff maths (base 2s, cap 60s,
 * ±1s jitter). Retries only on `RateLimitError`; any other error propagates
 * on the first throw.
 */

import { RateLimitError } from './errors.js';

export interface BackoffOptions {
  /** Max retry attempts after the first failure. Default 4. Set 0 to disable. */
  retries?: number;
  /** Base delay in ms (doubles each attempt). Default 2000. */
  baseMs?: number;
  /** Delay ceiling in ms. Default 60000. */
  capMs?: number;
  /** Max random jitter added to each delay, in ms. Default 1000. */
  jitterMs?: number;
  /** Called before each backoff sleep. */
  onRetry?: (info: { attempt: number; delayMs: number; error: RateLimitError }) => void;
  /** Injectable sleep (defaults to setTimeout). Handy in tests. */
  sleep?: (ms: number) => Promise<void>;
}

const defaultSleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/** Compute the jittered delay for a given 0-based attempt index. */
export function backoffDelayMs(attempt: number, opts: BackoffOptions = {}): number {
  const baseMs = opts.baseMs ?? 2_000;
  const capMs = opts.capMs ?? 60_000;
  const jitterMs = opts.jitterMs ?? 1_000;
  const delay = Math.min(baseMs * 2 ** attempt, capMs);
  return delay + Math.random() * jitterMs;
}

/**
 * Run `fn`, retrying on `RateLimitError` with exponential backoff. Re-throws
 * the last `RateLimitError` once `retries` is exhausted; any non-rate-limit
 * error is thrown immediately.
 */
export async function withBackoff<T>(
  fn: () => Promise<T>,
  opts: BackoffOptions = {},
): Promise<T> {
  const retries = opts.retries ?? 4;
  const sleep = opts.sleep ?? defaultSleep;

  let attempt = 0;
  for (;;) {
    try {
      return await fn();
    } catch (err) {
      if (!(err instanceof RateLimitError) || attempt >= retries) throw err;
      // Honor a server-supplied `Retry-After` when it asks for a longer wait
      // than our own backoff curve; never retry sooner than the server allows.
      const delayMs = Math.max(backoffDelayMs(attempt, opts), err.retryAfterMs ?? 0);
      opts.onRetry?.({ attempt: attempt + 1, delayMs, error: err });
      await sleep(delayMs);
      attempt++;
    }
  }
}
